fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePlanner.java

```diff
@@ -255,19 +260,79 @@ private LuceneQueryClause getQueryForAndOr(@Nonnull LucenePlanState state, @Nonn
             if (childMask != null) {
                 childMask.setSatisfied(true);
             }
-            childClauses.add(childClause);
+            if (childClause instanceof LuceneBooleanQuery && ((LuceneBooleanQuery)childClause).getOccur() == occur) {
+                childClauses.addAll(((LuceneBooleanQuery)childClause).getChildren());
+                if (childClause instanceof LuceneNotQuery) {
+                    negatedChildren.addAll(((LuceneNotQuery)childClause).getNegatedChildren());
+                }
+            } else {
+                childClauses.add(childClause);
+            }
         }
         if (filterMask != null && filterMask.getUnsatisfiedFilters().isEmpty()) {
             filterMask.setSatisfied(true);
         }
+        if (!negatedChildren.isEmpty()) {
+            return new LuceneNotQuery(childClauses, negatedChildren);
+        }
         // Don't do Lucene scan if none are satisfied, though.
         if (childClauses.isEmpty()) {
             return null;
         }
-        final BooleanClause.Occur occur = filter instanceof OrComponent ? BooleanClause.Occur.SHOULD : BooleanClause.Occur.MUST;
         return new LuceneBooleanQuery(childClauses, occur);
     }
 
+    @Nullable
+    private LuceneQueryClause getQueryForNot(@Nonnull LucenePlanState state, @Nonnull NotComponent filter,
+                                             @Nullable String parentFieldName, @Nullable FilterSatisfiedMask filterMask) {
+        final LuceneQueryClause childClause = getQueryForFilter(state, filter.getChild(), parentFieldName, filterMask == null ? null : filterMask.getChildren().get(0));
+        if (childClause == null) {
+            return null;
+        }
+        if (filterMask != null) {
+            filterMask.setSatisfied(true);
+        }
+        return negate(childClause);
+    }
+
+    @Nonnull
+    private static LuceneQueryClause negate(@Nonnull LuceneQueryClause clause) {
+        if (clause instanceof LuceneBooleanQuery) {
+            final LuceneBooleanQuery booleanQuery = (LuceneBooleanQuery)clause;
+            final BooleanClause.Occur occur = booleanQuery.getOccur();
+            if (occur == BooleanClause.Occur.MUST) {
+                List<LuceneQueryClause> clauses = new ArrayList<>();
+                for (LuceneQueryClause child : booleanQuery.getChildren()) {
+                    clauses.add(negate(child));
+                }
+                if (clause instanceof LuceneNotQuery) {
+                    LuceneNotQuery notQuery = (LuceneNotQuery)clause;
+                    if (clauses.isEmpty() && notQuery.getNegatedChildren().size() == 1) {
+                        return notQuery.getNegatedChildren().get(0);
+                    }
+                    clauses.addAll(notQuery.getNegatedChildren());
+                }
+                return new LuceneBooleanQuery(clauses, BooleanClause.Occur.SHOULD);
+            } else {
```

Comment:
Is there a missing case here for if `occur` is `SHOULD` instead of `MUST` or `MUST_NOT`? I'm a little worried that if we're not careful, this could result in accidentally including `(*:*)` in more queries than we'd like, but maybe that's a danger with negative queries more broadly

Reply:
A boolean query only has `SHOULD` or `MUST`. Those are the two branches of the `if`. `NOT` is not represented as one with `MUST_NOT`, but rather with as `MUST` with a set of negated children.

I think this can be made clearer here by using `switch`. I will do that.

More broadly, I tried a few approaches to reconcile the Boolean algebra trees that queries have with the Boolean queries with occurrence requirements that Lucene has. None of them were entirely satisfactory. One whose behavior was slightly more obvious was where the flattening and the addition of any `*:*` happened at `bind` time rather than plan time. In the end, though, I decided it was true to the intent of the API to have all such manipulation happen early and to have `bind` only be about deferring the concrete representation until the concrete values where known and no other rewriting.